server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/storage/XodusServerIssueStore.java

```diff
@@ -228,17 +226,47 @@ public void mergeIssues(String projectKey, String branchName, List<ServerIssue>
         issues.forEach(issue -> updateOrCreateIssue(fileEntity, issue, txn));
       });
       closedIssueKeysToDelete.forEach(issueKey -> remove(issueKey, txn));
-      branch.setProperty(LAST_SYNC_PROPERTY_NAME, syncTimestamp.toEpochMilli());
+      branch.setProperty(LAST_ISSUE_SYNC_PROPERTY_NAME, syncTimestamp.toEpochMilli());
+    }));
+  }
+
+  @Override
+  public void mergeTaintIssues(String projectKey, String branchName, List<ServerTaintIssue> issuesToMerge, Set<String> closedIssueKeysToDelete, Instant syncTimestamp) {
+    timed("Merged " + issuesToMerge.size() + " taint issues in store. Closed " + closedIssueKeysToDelete.size() + ".", () -> entityStore.executeInTransaction(txn -> {
+      var project = getOrCreateProject(projectKey, txn);
+      var branch = getOrCreateBranch(project, branchName, txn);
+      var issuesByFilePath = issuesToMerge.stream().collect(Collectors.groupingBy(ServerTaintIssue::getFilePath));
+      issuesByFilePath.forEach((filePath, issues) -> {
+        var fileEntity = getOrCreateFile(branch, filePath, txn);
+        issues.forEach(issue -> updateOrCreateTaintIssue(fileEntity, issue, txn));
+      });
+      closedIssueKeysToDelete.forEach(issueKey -> removeTaint(issueKey, txn));
+      branch.setProperty(LAST_TAINT_SYNC_PROPERTY_NAME, syncTimestamp.toEpochMilli());
     }));
   }
 
   @Override
-  public Optional<Instant> getLastSyncTimestamp(String projectKey, String branchName) {
+  public Optional<Instant> getLastIssueSyncTimestamp(String projectKey, String branchName) {
     return entityStore.computeInReadonlyTransaction(txn -> findUnique(txn, PROJECT_ENTITY_TYPE, KEY_PROPERTY_NAME, projectKey)
       .map(project -> project.getLinks(PROJECT_TO_BRANCHES_LINK_NAME))
       .flatMap(branches -> findUnique(txn, BRANCH_ENTITY_TYPE, NAME_PROPERTY_NAME, branchName))
       .map(branch -> {
-        var lastSyncMs = (Long) branch.getProperty(LAST_SYNC_PROPERTY_NAME);
+        var lastSyncMs = (Long) branch.getProperty(LAST_ISSUE_SYNC_PROPERTY_NAME);
+        if (lastSyncMs != null) {
+          return Optional.of(Instant.ofEpochMilli(lastSyncMs));
+        } else {
+          return Optional.<Instant>empty();
+        }
+      }).orElse(Optional.<Instant>empty()));
```

Comment:
I think we could avoid having those 2 Optional.empty, maybe something like (to be tested):

```
.map(branch -> branch.getProperty(LAST_ISSUE_SYNC_PROPERTY_NAME)
.map(lastSyncMs -> Instant.ofEpochMilli((Long) lastSyncMs)
```
same for the method below